### ISSUE TITLE
"unreachable" false positive for generator that emits no values

### ISSUE THREAD

**Issue by nathanielmanistaatgoogle** [PyCQA/pylint#1471]
Sometimes one needs to write a generator function that emits no values when iterated over:
```
def my_generator_function():
    <maybe do some stuff having side effects>
    return
    yield
```
. `pylint` (1.7.1) correctly detects that the `yield` statement is unreachable and emits an `unreachable` message, but I don't think it should emit that message in this case. The `yield` isn't there to be reached and executed; it's there to cause the behavior to be recognized as a generator rather than a function.

(Am I mistaken about an unreachable `yield` after `return` being the way to write a generator that emits no values?)

**Comment by rogalski** [comment]
On the other hand - is it a reachable code?

`yield` statement won't ever be reached, so flow analysis would say that it's unreachable, is as simple as that. `PyCharm` linter also warns about this case.

IMHO is such a corner case that it does not justify having a special case in code for that. I'd just disable this message and move on.

Although to be completely honest, I'm quite interested in hearing an input from other people :) .

**Comment by nathanielmanistaatgoogle** [comment]
At its most literal level of meaning, an `unreachable` message from `pylint` says that the `yield` is unreachable, and it is certainly true that the `yield` is unreachable, so it is literally correct.

But on a slightly more emotional (and I think more important) level the emission of a message from `pylint` says that "the code here is not as well-expressed as it could be; there is a change worth making to the code that caused the message to be emitted" and I think that's _not_ the case in this circumstance. The code has a perfectly fine expression; `pylint` shouldn't be acting in a way that would give authors impression that there's something to be changed.

Keep in mind that the opposite of emitting an `unreachable` message is "not emitting an `unreachable` message". It is not "emitting a `reachable` message". Changing `pylint` to refrain from emitting a message in this circumstance would not be the same as making a claim that the `yield` is executable.

Just to be clear about the circumstance itself: it's cases in which the unreachable code is only `yield` statements and only in behaviors that have no _reachable_ yield statements. (If a behavior has some reachable `yield` statements, then any unreachable `yield` statements should be cleaned up.)

I'm also interested in hearing from other folks.

**Comment by AWhetter** [comment]
Why not do something like

```python
def my_generator_like_function():
    # Do stuff
    return iter(())
```

Then the result of the function can be iterated and it's clearer that you want to iterate over nothing. Plus `()` is a constant so doesn't require construction time.

**Comment by PCManticore** [comment]
I do agree withb @AWhetter, since his suggestion seems way more clear in the intent than an unreachable yield. I do not think we should complicate this check.
